src/deploy/functions/runtimes/node/index.ts

```diff
@@ -66,13 +66,75 @@ export class Delegate {
   // to decide whether to use the JS export method of discovery or the HTTP container contract
   // method of discovery.
   _sdkVersion: string | undefined = undefined;
-  get sdkVersion() {
+  get sdkVersion(): string {
     if (this._sdkVersion === undefined) {
       this._sdkVersion = versioning.getFunctionsSDKVersion(this.sourceDir) || "";
     }
     return this._sdkVersion;
   }
 
+  _bin = "";
+  get bin(): string {
+    if (this._bin === "") {
+      this._bin = this.getNodeBinary();
+    }
+    return this._bin;
+  }
+
+  getNodeBinary(): string {
+    const requestedVersion = semver.coerce(this.runtime);
+    if (!requestedVersion) {
+      throw new FirebaseError(
+        `Could not determine version of the requested runtime: ${this.runtime}`
+      );
+    }
+    const hostVersion = process.versions.node;
+
+    const localNodePath = path.join(this.sourceDir, "node_modules/node");
+    const localNodeVersion = versioning.findModuleVersion("node", localNodePath);
+
+    if (localNodeVersion) {
+      if (semver.major(requestedVersion) === semver.major(localNodeVersion)) {
+        logLabeledSuccess(
+          "functions",
+          `Using node@${semver.major(localNodeVersion)} from local cache.`
+        );
+        return localNodePath;
+      }
+    }
+
+    // If the requested version is the same as the host, let's use that
+    if (semver.major(requestedVersion) === semver.major(hostVersion)) {
```

Comment:
Nit: reduce if-else stack:

```
if (semver.major(requestedVersion) === semver.major(hostVersion)) {
    logLabeledSuccess("functions", `Using node@${semver.major(hostVersion)} from host.`);
    return process.execPath;
}

if (!process.env.FIREPIT_VERSION) {
  logLabeledWarning(
          "functions",
          `Your requested "node" version "${semver.major(
            requestedVersion
          )}" doesn't match your global version "${semver.major(
            hostVersion
          )}". Using node@${semver.major(hostVersion)} from host.`
        );
    return process.execPath;
}


// Otherwise we'll warn and use the version that is currently running this process.
logLabeledWarning(
  "functions",
  `You've requested "node" version "${semver.major(
    requestedVersion
  )}", but the standalone Firebase CLI comes with bundled Node "${semver.major(
    hostVersion
  )}".`
);
logLabeledSuccess(
  "functions",
  `To use a different Node.js version, consider removing the standalone Firebase CLI and switching to "firebase-tools" on npm.`
);
return process.execPath;
```

Reply:
I like it! ❤️